src_Core/RISCY_OOO/procs/lib/IndexedMultiset.bsv

```diff
@@ -0,0 +1,230 @@
+                dynamicAssert(recsRead[index].count != 0, "removed empty item in indexed multiset");
+                removeW[i].wset(index);
+            endmethod
+        endinterface;
+    end
+
+    Bool anyEmpty = any(isEmpty,recsRead);
+    Bool anyFull = any(isFull,recsRead);
+    method ActionValue#(idxT) insert(datT d) if (anyEmpty && !anyFull);
+        let i <- findIdx(d,recsRead);
+        insertW.wset(IndexedMultisetInsert{idx:i, dat:d});
+        return i;
+    endmethod
+
+    method ActionValue#(IndexedMultisetIndices#(idxT)) insertAndReserve(datT ins, datT res);
+        let insIdx <- findIdx(ins,recsRead);
+        insertW.wset(IndexedMultisetInsert{idx:insIdx, dat:ins});
+        let newRecs = recsRead;
+        newRecs[insIdx].data = ins;
+        let resIdx <- findIdx(res,newRecs);
+        reserveW.wset(IndexedMultisetInsert{idx:resIdx, dat:res});
+        return IndexedMultisetIndices{inserted: insIdx, reserved: resIdx};
+    endmethod
+
+    method datT lookup(idxT i) = records[i].data;
+
+    interface rPort = removes;
+
+    method Action clearAll = clearW.send;
+endmodule
+
+// An implementation of Indexed Multiset that depends on inserting and removing in the same order.
+module mkIndexedMultisetQueue(IndexedMultiset#(Bit#(idxTSz), datT, remWidth))
+    provisos (
+        Alias#(idxT, Bit#(idxTSz)),
+        Bits#(datT, _datsz),
+        Eq#(datT)
+    );
+    Vector#(TExp#(idxTSz), Reg#(datT)) records <- replicateM(mkRegU);
+    Vector#(TExp#(idxTSz), datT) recsRead = readVReg(records);
+
+    Reg#(Bit#(TAdd#(idxTSz,1))) lhead <- mkReg(0);
+    Reg#(Bit#(TAdd#(idxTSz,1))) ltail <- mkReg(0);
+    idxT head = truncate(lhead);
+    Bit#(TAdd#(idxTSz,1)) level = lhead - ltail;
+    //Bool empty = (level==0);
+    Bool full  = (level==fromInteger(valueOf(TExp#(idxTSz))));
+    Bool almostFull = (level>=fromInteger(valueOf(TExp#(idxTSz)))-1);
+
+    RWire#(datT) insertW <- mkRWire;
+    RWire#(datT) reserveW <- mkRWire;
+    Vector#(remWidth, RWire#(idxT)) removeW <- replicateM(mkRWire);
+    PulseWire clearW <- mkPulseWire;
+
+    function Bit#(TAdd#(idxTSz,1)) updateWidePointer(Bit#(TAdd#(idxTSz,1)) lIdx, idxT idx);
+      idxT diff = idx - truncate(lIdx);
+      return lIdx + zeroExtend(diff);
+    endfunction
+    (* no_implicit_conditions *)
+    rule canon;
+        Vector#(TExp#(idxTSz),datT) recs = recsRead;
+        Bit#(TAdd#(idxTSz,1)) nlhead = lhead;
+        Bit#(TAdd#(idxTSz,1)) nltail = ltail;
+        if (insertW.wget matches tagged Valid .d) begin
+            $display("insert dat:%x nlhead:%d", {pack(d),12'b0}, nlhead);
+            idxT i = truncate(nlhead);
+            recs[i] = d;
+            nlhead = nlhead + 1;
+        end
+        if (reserveW.wget matches tagged Valid .d) begin
+            $display("reserve dat:%x nlhead:%d", {pack(d),12'b0}, nlhead);
+            idxT i = truncate(nlhead);
+            recs[i] = d;
+            //nlhead = nlhead + 1;
```

Comment:
Delete?

Reply:
Deleted all the debug output in this file for a new commt.
Not sure if you were suggesting that we don't write recs; this would almost work, as the next fetch (which must be to PPC and write to the same location, unless things are flushing) will likely write the same value to the same location again.  However I'm not sure if we can guarantee that this next fetch will always happen before this instruction gets to Decode, when it will need to decompress its PPC from that index.

Reply:
I meant the commented-out code